go/arrow/array/struct.go

```diff
@@ -47,21 +48,60 @@ func (a *Struct) Field(i int) Interface { return a.fields[i] }
 func (a *Struct) String() string {
 	o := new(strings.Builder)
 	o.WriteString("{")
+
+	structBitmap := a.NullBitmapBytes()
 	for i, v := range a.fields {
 		if i > 0 {
 			o.WriteString(" ")
 		}
+		if !bytes.Equal(structBitmap, v.NullBitmapBytes()) {
+			masked := newStructFieldWithParentValidityMask(a, i)
+			fmt.Fprintf(o, "%v", masked)
+			masked.Release()
+			continue
+		}
 		fmt.Fprintf(o, "%v", v)
 	}
 	o.WriteString("}")
 	return o.String()
 }
 
+// newStructFieldWithParentValidityMask returns the Interface at fieldIndex
+// with a nullBitmapBytes adjusted according on the parent struct nullBitmapBytes.
+// From the docs:
+//   "When reading the struct array the parent validity bitmap takes priority."
+func newStructFieldWithParentValidityMask(a *Struct, fieldIndex int) Interface {
```

Comment:
I thought about adding this to `Field(i)` but I decided against it because it would require unnecessary caching logic and I thought it was more in line with the way the other nested logical types implemented slicing.

https://github.com/apache/arrow/blob/a4a11bf838526a1b3f650616551f8a79e8db513b/go/arrow/array/list.go#L66
https://github.com/apache/arrow/blob/a4a11bf838526a1b3f650616551f8a79e8db513b/go/arrow/array/fixed_size_list.go#L66

Reply:
I moved it to a method on struct. https://github.com/apache/arrow/pull/5642/commits/4cfef3c61316c2644175b119d97d831c10926d50

Reply:
Ahh. I think I see what you're referring to with `Field(i)`. I guess we need to decide what the behavior of `Field(i)` should be. Should calling `Field(i)` return the masked field or the underling field (which could potentially have a different null bitmap)?

Reply:
I'm not sure if this is the desired behavior, but for what it's worth the Python version does not apply the struct null bitmap either.

```python
python
Python 3.7.3 | packaged by conda-forge | (default, Jul  1 2019, 14:38:56) 
[Clang 4.0.1 (tags/RELEASE_401/final)] :: Anaconda, Inc. on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import pyarrow as pa
>>> import numpy as np
>>> pa.__version__
'0.15.0'
>>> np.__version__
'1.17.3'
>>> ty = pa.struct([('x', pa.int16()),('y', pa.bool_())])
>>> xs = pa.array([5, None, 7], type=pa.int16())
>>> ys = pa.array([False, True, True])
>>> mask = pa.array(['a', 'b', 'c'], mask=np.array([True, False, False]))
>>> mask
<pyarrow.lib.StringArray object at 0x11995e108>
[
  null,
  "b",
  "c"
]
>>> arr = pa.StructArray.from_buffers(ty, 3, mask.buffers()[0:1], children=(xs, ys))
>>> arr
<pyarrow.lib.StructArray object at 0x11995e168>
-- is_valid:
  [
    false,
    true,
    true
  ]
-- child 0 type: int16
  [
    5,
    null,
    7
  ]
-- child 1 type: bool
  [
    false,
    true,
    true
  ]
>>> arr.field(0)
<pyarrow.lib.Int16Array object at 0x11995e1c8>
[
  5,
  null,
  7
]
>>> arr.field(0).null_count
1
>>> 
```

Reply:
The C++ implementation doesn't appear to try to apply the nulls on a call to `field()`: https://github.com/apache/arrow/blob/master/cpp/src/arrow/array.cc#L612

My vote would be to add a new method to Struct where the developer can choose to select a field with the nulls applied and leave the existing `Field` method as is.